### PR TITLE
Fix for default definition of Kokkos_ENABLE_PTHREAD

### PR DESCRIPTION
**Pull request by ndellingwood**

### CODE REVIEW

**Comment by ndellingwood** [comment]
When configuring with Trilinos we hit a case e.g. in Serial builds with `TPL_ENABLE_Pthread=ON` that resulted in incorrectly setting `Kokkos_ENABLE_PTHREAD=ON` within `KOKKOS_TPL_OPTION` cmake function by taking the TPL's default. In addition, a default `Kokkos_ENABLE_PTHREAD` value is needed in the cache by various packages in Trilinos

**Comment by ndellingwood** [comment]
Will cherry-pick to develop after review and PR is finalized

File: cmake/kokkos_enable_devices.cmake [hunk 1]
```diff
@@ -26,6 +26,9 @@ IF(Kokkos_ENABLE_PTHREAD)  # for backward compatibility
   LIST(APPEND KOKKOS_ENABLED_DEVICES THREADS)
   SET(KOKKOS_HAS_HOST ON)
   MESSAGE(DEPRECATION "The Kokkos_ENABLE_PTHREAD option is deprecated. Use Kokkos_ENABLE_THREADS instead!")
+ELSEIF(NOT Kokkos_ENABLE_THREADS)  # for backward compatibility
+  SET(Kokkos_ENABLE_PTHREAD OFF CACHE BOOL "Whether to build C++ threads backend")
+  SET(KOKKOS_ENABLE_PTHREAD OFF)
```
Comment (dalg24) [inline cmake/kokkos_enable_devices.cmake:31]: Please elaborate why this is needed

**Review by masterleinad — COMMENTED** [review]
If I understand correctly, the (main) problem is that `KOKKOS_ENABLE_(P)THREADS` is set whenever `PTHREAD` is enabled as a TPL in Trilinos and this behavior is not desired?

Wouldn't it make more sense to set `PTHREAD_DEFAULT=OFF` unconditionally instead (and not changing anything else)?

The current changes would remove `Kokkos_ENABLE_PTHREAD` as a cache variable if it is enabled as non-cache variable AFAICT.

File: cmake/kokkos_enable_devices.cmake (hunk 1 above)
Comment (ndellingwood) [reply cmake/kokkos_enable_devices.cmake:31]: This is to ensure that `{Kokkos,KOKKOS}_ENABLE_PTHREAD` has a default definition. Prior to the renaming to Threads, this was done by calling `KOKKOS_DEVICE_OPTION` with `PTHREAD`; currently this is done for `THREADS` (line 22 defaulted to OFF) with some follow-up if `Kokkos_ENABLE_PTHREAD` is already defined (lines 23-28), but there is no default "OFF" case for `Kokkos_ENABLE_PTHREAD`
This was needed with deprecated code enabled in some Trilinos packages (e.g. Tpetra) that check on the definition of `Kokkos_ENABLE_PTHREAD` during configuration

**Comment by ndellingwood** [comment]
> If I understand correctly, the (main) problem is that `KOKKOS_ENABLE_(P)THREADS` is set whenever `PTHREAD` is enabled as a TPL in Trilinos and this behavior is not desired?

That was part of the problem, where it seems the `KOKKOS_TPL_OPTION` takes the default from whether TriBITS has the TPL enabled (regardless of arguments to the routine). I can test again with `PTHREAD_DEFAULT=OFF` unconditionally as suggested but I think similar things I tried were still problematic

`Kokkos_ENABLE_PTHREAD` is still needed as a cache variable for checks by Trilinos packages when deprecated code is enabled, I attempted to deal with this with the changes in [cmake/kokkos_enable_devices.cmake](https://github.com/kokkos/kokkos/pull/4830/files/15da60732ce0f4f0efc98388bd51601b0142a83b#diff-b6ca20f17c15dd68998ee421461d8ebd0d4e37c3c681a628e16d523feb7f794f)

**Comment by masterleinad** [comment]
> `Kokkos_ENABLE_PTHREAD` is still needed as a cache variable for checks by Trilinos packages when deprecated code is enabled, I attempted to deal with this with the changes in [cmake/kokkos_enable_devices.cmake](https://github.com/kokkos/kokkos/pull/4830/files/15da60732ce0f4f0efc98388bd51601b0142a83b#diff-b6ca20f17c15dd68998ee421461d8ebd0d4e37c3c681a628e16d523feb7f794f)

I still see `Kokkos_ENABLE_PTHREAD` as a cache variable in the recent `develop` and the `kokkos-promotion` branch.

**Comment by ndellingwood** [comment]
> I still see `Kokkos_ENABLE_PTHREAD` as a cache variable in the recent `develop` and the `kokkos-promotion` branch.

Yes, without these changes and when using the configure setting -DTPL_ENABLE_Pthread=ON with Trilinos resulted in Kokkos_ENABLE_PTHREAD=ON on kokkos-promotion branch which caused problems (defined as OFF on the develop branch)

**Comment by masterleinad** [comment]
> Yes, without these changes and when using the configure setting -DTPL_ENABLE_Pthread=ON with Trilinos (Serial-only build) resulted in Kokkos_ENABLE_PTHREAD=ON on kokkos-promotion branch which caused problems (defined as OFF on the develop branch)

Yes, that's what I expected. I just try to understand why fixing this by setting the default to `OFF` is not sufficient as a fix. I thought you were saying that it wouldn't be a cache variable when you were saying
> Kokkos_ENABLE_PTHREAD is still needed as a cache variable for checks by Trilinos packages when deprecated code is enabled, I attempted to deal with this with the changes in [cmake/kokkos_enable_devices.cmake](https://github.com/kokkos/kokkos/pull/4830/files/15da60732ce0f4f0efc98388bd51601b0142a83b#diff-b6ca20f17c15dd68998ee421461d8ebd0d4e37c3c681a628e16d523feb7f794f)

**Comment by ndellingwood** [comment]
@masterleinad Yeah, there are a couple problems I'm trying to untangle in the PR making it messy to communicate.

You're right, the changes in [cmake/kokkos_enable_devices.cmake](https://github.com/kokkos/kokkos/pull/4830/files/15da60732ce0f4f0efc98388bd51601b0142a83b#diff-b6ca20f17c15dd68998ee421461d8ebd0d4e37c3c681a628e16d523feb7f794f) do set the Kokkos_ENABLE_PTHREAD cache variable.
Before I added that update, I made the changes in [cmake/kokkos_tpls.cmake](https://github.com/kokkos/kokkos/pull/4830/files#diff-f03402f2fdfc82f4841bf31404061645547a3839d82adda2ed0ef86cecc7fde6). In kokkos_tpls, `Kokkos_ENABLE_PTHREAD` was being set with preference to the default available through TriBITS which was ON due to the config setting TPL_ENABLE_Pthread=ON (seemingly regardless of `PTHREAD_DEFAULT`, TriBITS default is preferred when building with Trilinos). After removing the associated lines causing the error there is no default for `Kokkos_ENABLE_PTHREAD` unless adding something like those changes in kokkos_enable_device.cmake

I'm happy to make any changes if this should be handled differently, this PR got me past the problems I ran into

**Review by masterleinad — COMMENTED** [review]
I think I understand somewhat better now what's going on here. Before the change from pthreads to threads we had an
```
KOKKOS_DEVICE_OPTION(PTHREADS OFF HOST "Whether to build C++ threads backend")
```
option but also
```
KOKKOS_TPL_OPTION(PTHREAD ${PTHREAD_DEFAULT} TRIBITS Pthread)
```
We tried to rename the first one so that `Kokkos_ENABLE_PTHREADS` wouldn't be a (user-visible) cache variable but forgot to do anything about the latter option (so that it would still show up as cache variable).
Furthermore, setting the default for by defining `PTHREAD_DEFAULT` doesn't have any impact since we override it with the value of `TPL_ENABLE_Pthread`.

Hence, I think we should accept the changes in `cmake/kokkos_tpls.cmake`. It was our intent to get rid of the `cache` variable in the first place. Also, I would rather clean the places that need `Kokkos_ENABLE_PTHREAD` as cache variable in `Trilinos` and replace it with `Kokkos_ENABLE_THREADS`.
Only if that is not possible for whatever reason. I would rather have `Kokkos_ENABLE_PTHREAD` defines as hidden shared variable initialized with the value of `Kokkos_ENABLE_THREADS` using `mark_as_advanced`.

**Comment by ndellingwood** [comment]
@masterleinad thanks for the feedback. I'll update the PR to drop the cmake/kokkos_enable_devices.cmake changes and make the corresponding changes in the kokkos-promotion branch of Trilinos to replace `Kokkos_ENABLE_PTHREAD` with `Kokkos_ENABLE_THREADS`. I'll update this PR after testing to confirm I have things working properly

**Comment by ndellingwood** [comment]
I dropped the `cmake/kokkos_enable_devices.cmake` change. Trilinos changes to rename Kokkos_ENABLE_PTHREAD cache variable to Kokkos_ENABLE_THREADS are stored in my local kokkos-promotion branch, I'll push them up after updating the kokkos snapshot

**Review by dalg24 — APPROVED** [review]

**Pull request merged**